### ISSUE TITLE
Unable to write-psfmessage and include custom header information.

### ISSUE THREAD

**Issue by admn-monkey** [PowershellFrameworkCollective/psframework#567]
The following properly creates the expected logfile provider instance with the custom headers Bob, Sally, and Trevor. When using Write-PSFMessage, there doesn't appear to be any logical way to write to those columns. Calling to the provider without any special header info writes to the CSV as it should yet results in those special columns remaining empty. Nothing I do writes into those special headers. 

 

$paramSetPSFLoggingProvider = @{

                             Name          = 'logfile'

                             InstanceName  = $($ScriptName)

                             CsvDelimiter  = ';'

                             Headers       = 'Level','Tags','Message','Timestamp','ModuleName','Bob','Sally','Trevor'

                             FilePath      = "$($LogFilePath)\$($scriptName)-%Date%.csv"

                             Enabled       = $true

                             LogRotatePath = "$($LogFilePath)\$($scriptName)-*.csv"

                             }

             

Set-PSFLoggingProvider @paramSetPSFLoggingProvider

 The above works. The below does not and I am unable to ascertain how to accomplish the desired result. 

Write-PSFmessage -Message $($msg) -Level $($ID) -ModuleName $($ScriptName) -Bob No_Action -Tag SomeTag

 

Results in the error: "Write-PSFMessage : A parameter cannot be found that matches parameter name 'Bob'."

 

What is the correct way to create and use custom headers? I have exhausted my capabilities on searching and randomly testing. Since the Slack channel signup is broken too, I'm out of ideas.

**Comment by FriedrichWeinmann** [comment]
Hi,
this is not really an intended scenario for the logging, but it can be made to work at a performance cost:

```powershell
$paramSetPSFLoggingProvider = @{
	Name          = 'logfile'
	InstanceName  = $($ScriptName)
	CsvDelimiter  = ';'
	Headers       = 'Level', 'Tags', 'Message', 'Timestamp', 'ModuleName', @{N = 'Bob'; E = { $_.Data.Bob } }, @{N = 'Sally'; E = { $_.Data.Sally } }, @{N = 'Trevor'; E = { $_.Data.Trevor } }
	FilePath      = "$($LogFilePath)\$($scriptName)-%Date%.csv"
	Enabled       = $true
	LogRotatePath = "$($LogFilePath)\$($scriptName)-*.csv"
	Wait          = $true
}
Set-PSFLoggingProvider @paramSetPSFLoggingProvider

Write-PSFMessage -Message ($ID) -ModuleName $($ScriptName) -Tag SomeTag -Date @{ Bob = 'No_Action' }
```

**Comment by FriedrichWeinmann** [comment]
You can include the data field as a header, but CSV will not deal with it as desired.
Json format logs would handle the nesting as intended natively.

Write-PSFMessage does not support dynamic parameters for performance reasons, the `-Data` parameter is designed as an optional parameter logging providers may support, but are not required to. Some do so by default (for example eventlog), in case of logfile, csv is a flat table and thus unsuitable for nested data. The hack shown above is not ideal, but will do the job for CSV

**Comment by admn-monkey** [comment]
I'd first like to extend my sincere heartfelt thanks for the quick reply and solution. PSFramework is amazing and the effort you've put into it and continuing effort to update it is so sincerely appreciated. 

tldr: This is solved. I'm completely OK with the performance hit to get the ongoing results I'm after. One small ask; Are there any plans on the roadmap to include header functionality into logging to SQL? While there's certainly a dozen ways to send normalized data into SQL, the amazingly simple ability to just call Write-PSFMessage to another provider in a similar method would be incredibly useful. 

On to the long winded version... 

It's a data driven world and ultimately logging has moved well beyond error trapping. Frankly RFC 5424 needs a serious review, reconsideration, and rewrite since it was hardened way back in early 2009. While it's an understatement, I think everyone could agree the world has changed. 

I'm working with significant but delicate modifications within Active Directory as well as some future work in ADCS. Being able to selectively roll back is a key requirement for this type of work. Analysis, auditing, and trending are ultimately required too. Those last three aren't derived from CSV; I use SQL or other appropriate solutions where multiple data sets are more easily combined. Regarding trending, the ability to show results in an attractive graph to leadership is the icing that must always be included on the cake. I have to be able to standardize things like categorized logging so I can quickly adapt to the requirements for a given project.  

A few days ago I mentioned PSFramework to a friend who works in a completely separate company and industry. He's a Developer in a small / medium non-profit, so in many ways he's a one person band. Our data needs are significantly different yet he is interested in PSFramework as it has some incredibly attractive features. He uses Powershell to deliver fast diverse solutions and hold to reusable standards in the process. There's no need to reinvent the wheel if one doesn't have to and PSFramework beautifully fits into that niche area of data collection even though it was never intended to. 

Again, thanks for all the effort you've put into PSFramework; I know the team I work with will be using it. 

Below I've just corrected a typo or two for anyone else needing a similar solution. 


`

$paramSetPSFLoggingProvider = @{
	Name          = 'logfile'
	InstanceName  = $($ScriptName)
	CsvDelimiter  = ';'
	Headers       = 'Level', 'Tags', 'Message', 'Timestamp', 'ModuleName', @{N = 'Bob'; E = { $_.Data.Bob } }, @{N = 'Sally'; E = { $_.Data.Sally } }, @{N = 'Trevor'; E = { $_.Data.Trevor } }
	FilePath      = "$($LogFilePath)\$($scriptName)-%Date%.csv"
	Enabled       = $true
	LogRotatePath = "$($LogFilePath)\$($scriptName)-*.csv"
	Wait          = $true
}
Set-PSFLoggingProvider @paramSetPSFLoggingProvider

Write-PSFMessage -Message $($ID) -ModuleName $($ScriptName) -Tag SomeTag -Data @{ Bob = 'No_Action'; Sally = $($SomeVariable)}

`

**Comment by FriedrichWeinmann** [comment]
Heya,
I went ahead and made this better in the new 1.8.289 release:

```powershell
$paramSetPSFLoggingProvider = @{
	Name          = 'logfile'
	InstanceName  = $($ScriptName)
	CsvDelimiter  = ';'
	Headers       = 'Level', 'Tags', 'Message', 'Timestamp', 'ModuleName', 'Data.Bob as Bob', 'Data.Sally as Sally', 'Data.Trevor as Trevor'
	FilePath      = "$($LogFilePath)\$($scriptName)-%Date%.csv"
	Enabled       = $true
	LogRotatePath = "$($LogFilePath)\$($scriptName)-*.csv"
	Wait          = $true
}
Set-PSFLoggingProvider @paramSetPSFLoggingProvider

Write-PSFMessage -Message ($ID) -ModuleName $($ScriptName) -Tag SomeTag -Data @{ Bob = 'No_Action' }
```
